### ISSUE TITLE
Pointer tagging

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/backlog#91]
## What is pointer tagging?

I think on most (or quite possibly all) Oxc's supported 64-bit architectures, the top 6 or 7 bits of pointers are unused, and could be used to pack additional data into the pointer's bytes ("pointer tagging").

One useful article on this subject: https://muxup.com/2023q4/storing-data-in-pointers

## What could we do with it?

Potentially we could squeeze `Expression` and `Statement` down from 16 to 8 bytes, by storing the enum discriminant in top 6/7 bits.

`Expression` and `Statement` are both extremely common in AST, and also are often passed around between functions. So squeezing their size might make a significant difference to performance.

## Difficulties

However, there are challenges:

1. Platform support

* Need to define which platforms we support.
* Need to make sure all those platforms support pointer tagging.
* Need to test on those platforms.

2. Enum matching

* How to support `match` on enums where our memory representation of enums does not match Rust's native representation?
* There may be ways to do this and preserve an ergonomic interface, but it's unclear.

## What to do for now?

There are major challenges to implementing pointer tagging. So unlikely that we'd want to attempt this now.

But I've opened this issue to have as a reference, and as a reminder that we may not want to use the high bits of pointers to pack *other* data, to leave the door open for the enum optimization in future.

## Are there alternatives?

Yes. We could also reduce the size of AST types by more efficiently storing the data backing enums e.g.:

### `Vec<T>`

Store enum discriminants and enum "payloads" separately in e.g. `Vec<Statement>`.

Essentially it'd be an SoA pair of `Vec`s - `Vec<u8>` (discriminants) + `Vec<NonNull<()>>` (pointers) = 9 bytes per element, vs current 16 bytes.

### Structs

Storing enum discriminants in structs separately from the "payloads". e.g. `BinaryExpression` would be stored as:

```rs
struct BinaryExpressionRepr {
    span: Span,
    left_ptr: NonNull<()>,
    right_ptr: NonNull<()>,
    left_discriminant: u8,
    right_discriminant: u8,
    operator: BinaryOperator,
}
```

This is 32 bytes (vs 48 bytes at present) - and still has 5 padding bytes which could fit a `u32` Node ID.

**Comment by DonIsaac** [comment]
Just a thing to keep in mind: Our [Wasm32 bindings](https://www.npmjs.com/package/@oxc-resolver/binding-wasm32-wasi) are quite popular; I don't think we could drop support for 32 bit architectures.

![image](https://github.com/user-attachments/assets/1527d2b2-7d18-46dd-9ddc-012b1681ccaf)

**Comment by overlookmotel** [comment]
Yes, we'd need to keep things as they are for 32-bit archs (which I think in practice is just WASM). This optimization is only possible with 64-bit pointers.

**Comment by overlookmotel** [comment]
[Article on Bun blog](https://bun.sh/blog/how-bun-supports-v8-apis-without-using-v8-part-1#jsvalue) mentions that all 64-bit platforms they support only use 48-bit address space, leaving the top 16 bits of pointers free for tagging. We haven't defined what platforms we support yet, but I imagine we could do the same as Bun.

**Comment by overlookmotel** [comment]
I've come up with an API which I think is reasonably ergonomic.

`Expression` is used as example here, but we'd do the same for all AST enum types e.g. `Statement`, `Declaration`, `PropertyKey`.

On 64-bit platforms, store enum discriminant in top 7 bits of pointer. This reduces size of enums from 16 bytes to 8 bytes.

`Compact<Expression>` is a compact representation of `Expression`, which uses this compression scheme.

## API changes overview

Main changes to how we interact with AST nodes:

* Everywhere we hold or pass `Expression`s, pass `Compact<Expression>` instead (and same for all other enum types).
* `AstBuilder` methods which build enums return `Compact<T>` not `T`.
* To `match` on a `Compact<Expression>`, convert to `ExpressionRef` or `ExpressionMut` first. i.e. `match expr { ... }` -> `match expr.as_ref() { ... }`.

Full examples of API usage below.

## AST types

Enum types remain same as before:

```rs
pub enum Expression<'a> {
    BooleanLiteral(Box<'a, BooleanLiteral>),
    NullLiteral(Box<'a, NullLiteral>),
    NumericLiteral(Box<'a, NumericLiteral<'a>>),
    // ...
}
```

Structs contain `Compact<T>` instead of `T` for enum types:

```rs
pub struct BinaryExpression<'a> {
    pub span: Span,
    pub left: Compact<Expression<'a>>,
    pub operator: BinaryOperator,
    pub right: Compact<Expression<'a>>,
}
```

`Compact<Expression>` is 8 bytes vs `Expression` 16 bytes. So `BinaryExpression` drops from 48 bytes to 32 bytes.

## Utility types

```rs
pub struct Compact<T> {
    ptr: NonNull<u64>,
    _marker: PhantomData<T>,
}
```

Codegen `*Type`, `*Ref` and `*Mut` types for all enum AST types:

```rs
#[repr(u8)]
pub enum ExpressionType {
    BooleanLiteral,
    NullLiteral,
    NumericLiteral,
    // ...
}

pub enum ExpressionRef<'a, 'r> {
    BooleanLiteral(&'r BooleanLiteral),
    NullLiteral(&'r NullLiteral),
    NumericLiteral(&'r NumericLiteral<'a>),
    // ...
}

pub enum ExpressionMut<'a, 'r> {
    BooleanLiteral(&'r mut BooleanLiteral),
    NullLiteral(&'r mut NullLiteral),
    NumericLiteral(&'r mut NumericLiteral<'a>),
    // ...
}
```

## Usage examples

```rs
fn use_immutable_ref<'a>(expr: &Compact<Expression<'a>>) {
    // Test type of `Compact<Expression>`
    let is_binary = expr.is_binary_expression();
    let is_bool_or_number = matches!(expr.ty(), ExpressionType::BooleanLiteral | ExpressionType::NumericLiteral);

    // Use `Expression`'s methods via `ExpressionRef`
    let is_ts = expr.as_ref().is_typescript_syntax();

    // Match a `&Compact<Expression>` via `ExpressionRef`
    match expr.as_ref() { // <-- `.as_ref()` is the addition here
        ExpressionRef::BinaryExpression(bin_expr) => {
            // `bin_expr` is a `&BinaryExpression`
        }
        ExpressionRef::UnaryExpression(unary_expr) => {
            // `unary_expr` is a `&UnaryExpression`
        }
        _ => {}
    }
}

fn use_mutable_ref<'a>(expr: &mut Compact<Expression<'a>>, ctx: &mut TraverseCtx<'a>) {
    // Test type of `Compact<Expression>`
    let is_binary = expr.is_binary_expression();
    let is_bool_or_number = matches!(expr.ty(), ExpressionType::BooleanLiteral | ExpressionType::NumericLiteral);

    // Use `Expression`'s methods via `ExpressionMut`
    *expr.as_mut().span_mut() = SPAN;

    // Match a `&mut Compact<Expression>` via `ExpressionMut`
    match expr.as_mut() { // <-- `.as_mut()` is the addition here
        ExpressionMut::BinaryExpression(bin_expr) => {
            // `bin_expr` is a `&mut BinaryExpression`
        }
        ExpressionMut::UnaryExpression(unary_expr) => {
            // `unary_expr` is a `&mut UnaryExpression`
        }
        _ => {}
    }

    // Replace node
    *expr = ctx.ast.expression_null(SPAN);
}

fn use_owned<'a>(expr: Compact<Expression<'a>>) -> Compact<Expression<'a>> {
    // Unwrap to `Expression`
    let expr = expr.into_inner();

    // Match `Expression` in usual way
    match &expr {
        Expression::BinaryExpression(bin_expr) => {
            // `bin_expr` is a `&Box<'a, BinaryExpression<'a>>`
        }
        Expression::UnaryExpression(unary_expr) => {
            // `unary_expr` is a `&Box<'a, UnaryExpression<'a>>`
        }
    }

    // Wrap in `Compact` again
    let expr = expr.into_compact();

    expr
}
```

## Implementation

Codegen conversion methods for AST enum types as below.

Note: All these methods will be boiled down by compiler to 3 instructions or less. Pointer maths uses bitwise and/or/shift, which are the cheapest of all instructions.

```rs
impl<'a> Compact<Expression<'a>> {
    pub fn ty(&self) -> ExpressionType {
        unsafe { std::mem::transmute(self.tag()) }
    }

    pub fn as_ref<'r>(&'r self) -> ExpressionRef<'a, 'r> {
        unsafe {
            match self.ty() {
                ExpressionType::BooleanLiteral => {
                    ExpressionRef::BooleanLiteral(self.payload().cast::<BooleanLiteral>().as_ref())
                }
                ExpressionType::NullLiteral => {
                    ExpressionRef::NullLiteral(self.payload().cast::<NullLiteral>().as_ref())
                }
                ExpressionType::NumericLiteral => {
                    ExpressionRef::NumericLiteral(self.payload().cast::<NumericLiteral>().as_ref())
                }
                // ...etc...
            }
        }
    }

    pub fn as_mut<'r>(&'r mut self) -> ExpressionMut<'a, 'r> {
        unsafe {
            match self.ty() {
                ExpressionType::BooleanLiteral => {
                    ExpressionMut::BooleanLiteral(self.payload().cast::<BooleanLiteral>().as_mut())
                }
                ExpressionType::NullLiteral => {
                    ExpressionMut::NullLiteral(self.payload().cast::<NullLiteral>().as_mut())
                }
                ExpressionType::NumericLiteral => {
                    ExpressionMut::NumericLiteral(self.payload().cast::<NumericLiteral>().as_mut())
                }
                // ...etc...
            }
        }
    }

    pub fn into_inner(self) -> Expression<'a> {
        unsafe {
            match self.ty() {
                ExpressionType::BooleanLiteral => {
                    Expression::BooleanLiteral(Box::from_non_null(self.payload().cast::<BooleanLiteral>()))
                }
                ExpressionType::NullLiteral => {
                    Expression::NullLiteral(Box::from_non_null(self.payload().cast::<NullLiteral>()))
                }
                ExpressionType::NumericLiteral => {
                    Expression::NumericLiteral(Box::from_non_null(self.payload().cast::<NumericLiteral>()))
                }
                // ...etc...
            }
        }
    }

    pub fn is_boolean_literal(&self) -> bool {
        self.ty() == ExpressionType::BooleanLiteral
    }
    // ...etc...
}

impl<'a> Expression<'a> {
    pub fn into_compact(self) -> Compact<Expression<'a>> {
        unsafe {
            match self {
                Expression::BooleanLiteral(boxed) => Compact::new(ExpressionType::BooleanLiteral as u8, boxed),
                Expression::NullLiteral(boxed) => Compact::new(ExpressionType::NullLiteral as u8, boxed),
                Expression::NumericLiteral(boxed) => Compact::new(ExpressionType::NumericLiteral as u8, boxed),
                // ...
            }
        }
    }
}

impl<'a> From<Expression<'a>> for Compact<Expression<'a>> {
    fn from(node: Expression<'a>) -> Self {
        node.into_compact()
    }
}

impl<'a> From<Compact<Expression<'a>>> for Expression<'a> {
    fn from(compact: Compact<Expression<'a>>) -> Self {
        compact.into_inner()
    }
}

// `Deref` impls mean you can use `Expression`'s methods on an `ExpressionRef` or `ExpressionMut`.
// e.g. `compact_expr.as_ref().span()`, `compact_expr.as_ref().is_typescript_syntax()`.
// Note: I don't think it's possible to implement `Deref` on `Compact<Expression>` directly.
impl<'a, 'r> Deref for ExpressionRef<'a, 'r> {
    type Target = Expression<'a>;
    fn deref(&self) -> &Self::Target {
        unsafe { NonNull::from_ref(self).byte_add(8).cast::<Self::Target>().as_ref() }
    }
}

impl<'a, 'r> Deref for ExpressionMut<'a, 'r> {
    type Target = Expression<'a>;
    fn deref(&self) -> &Self::Target {
        unsafe { NonNull::from_ref(self).byte_add(8).cast::<Self::Target>().as_ref() }
    }
}

// `*compact_expr.as_mut() = Expression::NullLiteral(...)` wouldn't do anything, because it'd only
// mutate the `ExpressionMut`, not the underlying `Compact<Expression>`.
// However, mutating properties of the `Expression` would work, as would calling methods of `Expression`
// which take a `&mut self` e.g. `compact_expr.as_mut().set_span(SPAN)`.
impl<'a, 'r> DerefMut for ExpressionMut<'a, 'r> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        unsafe { NonNull::from_ref(self).byte_add(8).cast::<Self::Target>().as_mut() }
    }
}
```

```rs
impl<T: Sized> Compact<T> {
    unsafe fn new<P>(ty: u8, boxed: Box<'_, P>) -> Self {
        let ptr = boxed.into_non_null().as_ptr().cast::<u64>();
        let tagged = ptr as usize | ((ty as usize) << 57);
        let ptr = ptr.wrapping_byte_add(tagged - ptr as usize);
        let ptr = unsafe { NonNull::new_unchecked(ptr) };
        Self { ptr, _marker: PhantomData }
    }

    fn tag(&self) -> u8 {
        ((self.ptr.as_ptr() as usize) >> 57) as u8
    }

    fn payload(&self) -> NonNull<u64> {
        let ptr = self.ptr.as_ptr().wrapping_byte_sub((self.tag() as usize) << 57);
        unsafe { NonNull::new_unchecked(ptr) }
    }
}
```

## Platform compatibility

We're only using top 7 bits of pointer for the tag. This retains compatibility even with obscure platforms/configurations which use [5-level page tables](https://en.wikipedia.org/wiki/Intel_5-level_paging) and a 57-bit virtual address space.

It seems very likely that we will not run into compatibility issues on any 64-bit platform either now or in the future. 57-bit address space allows up to 128 *petabytes*. It's hard to imagine systems requiring more RAM than that in the conceivable future.

Note: This is more conservative than Bun, which is using pointer tagging with 16-bit tags.

### WASM

Tagging is not viable on 32-bit platforms (WASM). `Compact<T>` would just be a wrapper around `T` on WASM (8 bytes).

We could easily also implement in a way where memory representations of an AST on a 64-bit system and on 32-bit WASM are identical, so AST becomes portable between native code and WASM without any serialization/deserialization.

## Conclusion

It's hard to know without trying it, but I expect this could have significant positive performance impact:

* `Expression` and `Statement` are the most common nodes in AST. Halving their size will reduce size of AST by (at a guess) around 20%. This should produce better cache performance, and lower memory usage.
* When a `Vec<Statement>` grows, there are half as many bytes to copy.
* `Expression` now fits into a single register (instead of 2 registers). So more work can happen directly in registers, resulting in less memory operations. Everywhere we pass around owned `Expression`s, `Statement`s etc (e.g. parser) will benefit from this.

While the API does add some boilerplate (additional `.as_ref()`, `.as_mut()`), personally I don't think it's *too* bad. If the perf boost is significant, I think it would be worth it.

So, in my opinion, I think we should try this.

**Comment by overlookmotel** [comment]
Actually it may be even better to shift pointer left by 7 bits, and store tag in *bottom* 7 bits. The assembly is slightly tighter, and it performs sign extension on the pointer when unpacking, which extends support to pointers to top half of virtual memory (stack). Sign extension with the "tag in high bits" scheme requires more instructions.

https://godbolt.org/z/9fYGc1nxf

**Comment by overlookmotel** [comment]
If the types of all enum variants are aligned on at least 2 (which all our AST types are), that guarantees that lowest bit of pointer is 0.

Therefore, if shift the pointer left by 7 bits, and store tag in bottom 7 bits, the tag can be read just by reading the lowest byte, without masking it `& 127`. This shaves 1 instruction off getting the tag (so now it's just a single operation).

https://godbolt.org/z/zEahrrfea

Interestingly, the above Godbolt example shows that compiler seems to prefer ADD instruction over OR when combining tag and pointer, if it knows the tag is <= 127. I'm not sure if that's ideal or not. I would have assumed OR was preferable. But probably the compiler knows better than me!

**Comment by overlookmotel** [comment]
[`with_addr`](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) and [`map_addr`](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) pointer methods are now stable on our MSRV.

I don't *think* this gives any perf advantage over `wrapping_add` / `wrapping_sub`. But it would help make the code for pointer tagging simpler and more understandable.

**Comment by overlookmotel** [comment]
https://github.com/oxc-project/oxc/pull/10884 suggests a 2%-3% reduction in memory usage in arena translates to ~1% perf improvement across the board.

Some rough calculations suggested that halving the size of `Expression`, `Statement`, etc would reduce size of AST by 15%-20%. So we could expect maybe 5%-10% perf improvement from this change due to lower memory usage / better L1 cache utilization.

The other gains mentioned at bottom of https://github.com/oxc-project/backlog/issues/91#issuecomment-2585286030 would be on top of that. So potentially it could be quite good!

**Comment by overlookmotel** [comment]
Partial implementation on [overlookmotel/compact-pointer-tagging](https://github.com/oxc-project/oxc/tree/overlookmotel/compact-pointer-tagging) branch.

**Comment by overlookmotel** [comment]
https://github.com/oxc-project/oxc/pull/20489 demonstrated good perf and memory gains, but huge code churn.

To reduce the churn, I've added an [alternative implementation](https://github.com/oxc-project/oxc/blob/bc791e4134b276e01be6fbca9110833c512b8bee/crates/oxc_ast/src/compact_macro.rs) which provides a macro for matching (on [overlookmotel/compact-pointer-tagging](https://github.com/oxc-project/oxc/tree/overlookmotel/compact-pointer-tagging) branch, along with original impl).

It's a declarative macro, __not__ a proc macro, and a single macro handles all cases, regardless of whether the matched expression is an owned `T`, or `&T` / `&mut T` reference.

Without pointer tagging:

```rs
match expr {
  Expression::BinaryExpression(e) => process_bin_expr(e),
  Expression::UnaryExpression(e) => process_unary_expr(e),
  _ => {}
}
```

With pointer tagging and macro:

```rs
match_compact!(expr {
  Expression::BinaryExpression(e) => process_bin_expr(e),
  Expression::UnaryExpression(e) => process_unary_expr(e),
  _ => {}
})
```

Whereas, with original proposal above:

```rs
match expr.as_ref() {
  ExpressionRef::BinaryExpression(e) => process_bin_expr(e),
  ExpressionRef::UnaryExpression(e) => process_unary_expr(e),
  _ => {}
}
```

Macros have their downsides, but in this case I think there's an argument for it - reads clearer, and minimal diff vs the original code - only first and last lines change, the match body stays exactly the same as it was before.

**Comment by Boshen** [comment]
Still gross :-)

**Comment by overlookmotel** [comment]
Yeah, there's no way to avoid some code modifications if we want the perf + memory gain. Rust doesn't have 1st-class support for pointer tagging.
